MetadataExtractor/IO/BufferReader.cs

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) Drew Noakes and contributors. All Rights Reserved. Licensed under the Apache License, Version 2.0. See LICENSE in the project root for license information.
+
+using System.Buffers.Binary;
+
+namespace MetadataExtractor.IO;
+
+internal ref struct BufferReader(ReadOnlySpan<byte> bytes, bool isBigEndian)
+{
+    private readonly ReadOnlySpan<byte> _bytes = bytes;
+    private int _position = 0;
+    private bool _isBigEndian = isBigEndian;
+
+    public long Position => _position;
+
+    public byte GetByte()
+    {
+        if (_position >= _bytes.Length)
+            throw new IOException("End of data reached.");
+
+        return _bytes[_position++];
+    }
+
+    public void GetBytes(scoped Span<byte> bytes)
+    {
+        if (_position + bytes.Length > _bytes.Length)
+            throw new IOException("End of data reached.");
+
+        _bytes.Slice(_position, bytes.Length).CopyTo(bytes);
+        _position += bytes.Length;
+    }
+
+    public byte[] GetBytes(int count)
+    {
+        if (_position + count > _bytes.Length)
+            throw new IOException("End of data reached.");
+
+        var bytes = new byte[count];
+
+        _bytes.Slice(0, count).CopyTo(bytes);
+        _position += count;
+        return bytes;
+    }
+
+    public int Available()
+    {
+        return _bytes.Length - _position;
+    }
+
+    public void Skip(long n)
+    {
+        if (n < 0)
+            throw new ArgumentException("n must be zero or greater.");
+
+        if (_position + n > _bytes.Length)
+            throw new IOException("End of data reached.");
+
+        _position += unchecked((int)n);
```

Comment:
Maybe checked would be better here. What do you think?

Reply:
That would be safer, yes.  Let me spent some time on this file, and go through usual diligence.  I was fairly haphazard with copying / pasting here, and relying on test coverage.

Reply:
I generally rely upon the regression test suite to validate changes, as it flexes loads more code paths than unit tests do. That's how I found the regressions fixed in #391. Do you have that repo cloned locally? Let me know if you have any questions about getting that running locally. There are some docs at https://github.com/drewnoakes/metadata-extractor/wiki/Working-with-test-images but they might be out of date. The source to run is actually in the `metadata-extractor-images` repo. If you don't have the Java version built locally, you can comment out the `JavaRunner` line and rely upon the files already present in the repo. Note that the output files will vary based on Java/.NET version. In general, the `main` branch should produce the output in that repo (unless I forget to push).